android/src/main/kotlin/co/doneservices/callkeep/CallkeepSoundPlayerService.kt

```diff
@@ -0,0 +1,151 @@
+package co.doneservices.callkeep
+
+import android.app.Service
+import android.content.Context
+import android.content.Intent
+import android.media.AudioAttributes
+import android.media.AudioManager
+import android.media.MediaPlayer
+import android.media.RingtoneManager
+import android.net.Uri
+import android.os.*
+import android.text.TextUtils
+
+class CallKeepSoundPlayerService : Service() {
+
+    private var vibrator: Vibrator? = null
+    private var audioManager: AudioManager? = null
+
+    private var mediaPlayer: MediaPlayer? = null
+    private var data: Bundle? = null
+
+    override fun onBind(p0: Intent?): IBinder? {
+        return null
+    }
+
+    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
+        this.prepare()
+        this.playSound(intent)
+        this.playVibrator()
+        return START_STICKY;
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        mediaPlayer?.stop()
+        mediaPlayer?.release()
+        vibrator?.cancel()
+    }
+
+    private fun prepare() {
+        mediaPlayer?.stop()
+        mediaPlayer?.release()
+        vibrator?.cancel()
+    }
+
+    private fun playVibrator() {
+        vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+            val vibratorManager = this.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
+            vibratorManager.defaultVibrator
+        } else {
+            getSystemService(VIBRATOR_SERVICE) as Vibrator
+        }
+        audioManager = this.getSystemService(AUDIO_SERVICE) as AudioManager
+        when (audioManager?.ringerMode) {
+            AudioManager.RINGER_MODE_SILENT -> {
+            }
+            else -> {
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                    vibrator?.vibrate(VibrationEffect.createWaveform(longArrayOf(0L, 1000L, 1000L), 0))
+                } else {
+                    vibrator?.vibrate(longArrayOf(0L, 1000L, 1000L), 0)
+                }
+            }
+        }
+    }
+
+    private fun playSound(intent: Intent?) {
+        this.data = intent?.extras
+        val sound = this.data?.getString(
+                CallKeepBroadcastReceiver.EXTRA_CALLKEEP_RINGTONE_FILE_NAME,
+                ""
+        )
+        var uri = sound?.let { getRingtoneUri(it) }
+        if (uri == null) {
+            uri = RingtoneManager.getActualDefaultRingtoneUri(
+                    this@CallKeepSoundPlayerService,
+                    RingtoneManager.TYPE_RINGTONE
+            )
+        }
+        try {
+            mediaPlayer(uri!!)
+        } catch (e: Exception) {
+            try {
+                uri = getRingtoneUri("ringtone_default")
+                mediaPlayer(uri!!)
+            } catch (e2: Exception) {
+                e2.printStackTrace()
+            }
+        }
+    }
+
+    private fun mediaPlayer(uri: Uri) {
```

Comment:
Shouldn't this function be called something like `playLoopingSound`? Right now it's `mediaPlayer` but it doesn't return a media player. Rather it sets one up and makes it start playing a sound.

Reply:
Also how do you make the player stop playing later? If you never get the reference to the player.

Reply:
I agree on the naming part.
As for playing and stopping, this is an Android service so we usually play it by starting the service and stop it by stopping the service.
```
val soundPlayerServiceIntent =
Intent(context, CallKeepSoundPlayerService::class.java)
soundPlayerServiceIntent.putExtras(data)
context.startService(soundPlayerServiceIntent)
```

```
context.stopService(Intent(context, CallKeepSoundPlayerService::class.java))
```